yarpl/include/yarpl/flowable/Subscriber.h

```diff
@@ -15,40 +15,129 @@ namespace flowable {
 template <typename T>
 class Subscriber : public virtual Refcounted, public yarpl::enable_get_ref {
  public:
+  virtual void onSubscribe(Reference<Subscription>) = 0;
+  virtual void onComplete() = 0;
+  virtual void onError(folly::exception_wrapper) = 0;
+  virtual void onNext(T) = 0;
+};
+
+// codemod all things that inherit from Subscriber to inherit from
+// InternalSubscriber
+template <typename T>
+class InternalSubscriber : public Subscriber<T> {
+ public:
   // Note: If any of the following methods is overridden in a subclass, the new
   // methods SHOULD ensure that these are invoked as well.
-  virtual void onSubscribe(Reference<Subscription> subscription) {
+  void onSubscribe(Reference<Subscription> subscription) override {
     DCHECK(subscription);
-
-    if (subscription_) {
-      subscription->cancel();
-      return;
-    }
-
+    CHECK(!subscription_);
     subscription_ = std::move(subscription);
   }
 
   // No further calls to the subscription after this method is invoked.
-  virtual void onComplete() {
+  void onComplete() override {
     DCHECK(subscription_) << "Calling onComplete() without a subscription";
     subscription_.reset();
   }
 
   // No further calls to the subscription after this method is invoked.
-  virtual void onError(folly::exception_wrapper) {
+  void onError(folly::exception_wrapper) override {
     DCHECK(subscription_) << "Calling onError() without a subscription";
     subscription_.reset();
   }
 
-  virtual void onNext(T) = 0;
-
  protected:
-  Subscription* subscription() {
-    return subscription_.operator->();
+  Reference<Subscription> subscription() {
+    return subscription_;
   }
 
  private:
   Reference<Subscription> subscription_;
 };
+
+// T : Type of Flowable that this Subscriber operates on
+// keeps_reference_to_this : Indicates that the subclass will ensure
+// that a reference to `this` will stay live as long as `this` can
+// be accessed (eg understands that it's possible to drop all references
+// to the object within a method of that object)
+// Classes that ensure this can set keeps_reference_to_this to `true`
+// as an optimization (to prevent an atomic inc/dec pair)
+#define SUBSCRIBER_KEEP_SELF() \
+  Reference<SafeSubscriber> self; \
+  if (!keeps_reference_to_this) { \
+    self = this->ref_from_this(this); \
+  }
+
+template <typename T, bool keeps_reference_to_this = false>
```

Comment:
I don't see any uses of keeps_reference_to_this being any different than default false.
Can we discuss it in a separate PR? It doesn't seem to be essential to the SubscriberBase.